Add wp-parsely loader information to SDS

jesse-greathouse opened the pull request:
## Description

This adds wp-parsely plugin-related information to the site details index, with the helper of a new `Parsely_Loader_Info` class that is introduced to help keep track of how (and if) the plugin is being loaded.

Additionally:

- Avoid fatal errors when activating a self-installed version of the wp-parsely plugin whilst the mu-plugins version was already active.
- Cleanup loader logic, so that we can more effectively take into account the `wpvip_parsely_load_mu` filter & `_wpvip_parsely_mu` option.
- Move `maybe_load_plugin()` logic to the later `plugins_loaded` hook priority 1, rather than `muplugins_loaded`. This allows us to safely ensure a self-installed version of the plugin is not being loaded.

## Changelog Description
Information about how a site uses the parse.ly plugin has been added to the Site Details Index. The bundled version of wp-parsely plugin now loads on the `plugins_loaded` hook rather than on `muplugins_loaded` when enabled.

## Checklist

Please make sure the items below have been covered before requesting a review:

- [x] This change works and has been tested locally (or has an appropriate fallback).
- [ ] This change works and has been tested on a Go sandbox.
- [ ] This change has relevant unit tests (if applicable).
- [ ] This change has relevant documentation additions / updates (if applicable).
- [x] I've created a changelog description that aligns with the provided examples.

## Steps to test

Try out the various states, ensure all is loading correctly & reporting the correct loader information.

- Parsely not enabled.
- Parsely enabled by the option.
- Parsely enabled by the filter.
- Parsely enabled via a self-installed version.
- Parsely enabled by the option but disabled by the filter.

jesse-greathouse committed:
Removing jetpack-11.2 for subtree replacement to 11.2

jesse-greathouse committed:
Update jetpack 11.2 subtree to tag 11.2

jesse-greathouse committed:
Squashed 'jetpack-11.2/' content from commit a8dd6ecff1

git-subtree-dir: jetpack-11.2
git-subtree-split: a8dd6ecff13fba01eae7fd86088f93862425d0a6

jesse-greathouse committed:
Merge branch 'develop' of github.com:Automattic/vip-go-mu-plugins into update/SDS-parsely-plugin-stats

jesse-greathouse committed:
parsely details enhanced

jesse-greathouse committed:
added constants and get_parsely_info

jesse-greathouse committed:
Add get_parsely_info method to set_env_and_core method to produce site_details object in config/class-site-details-index.php

jesse-greathouse committed:
added missing docblocks

rebeccahum commented on config/class-site-details-index.php:
```diff
@@ -228,6 +252,100 @@ public function get_jetpack_info() {
+	 */
+	public function get_parsely_service_type( $parsely ) {
+		if ( is_array( $parsely ) && isset( $parsely['apikey'] ) && '' !== $parsely['apikey'] ) {
+			$result = $this->call_parsely_api( $parsely );
```
Leaving for historical purposes — there were concerns about this API call flagged on Slack by @WPprodigy

acicovic commented:
I've opened a Slack thread about this PR.

jesse-greathouse committed:
Merge branch 'develop' into update/SDS-parsely-plugin-stats

jesse-greathouse committed:
Changed get_parsely_version to revert to 'NONE' if there was no Parse.ly version detected.
Changed the methodology for detecting a hidden installation in get_parsely_integration_type
Changed the parsely_service_type logic to infer PAID when either mu-plugins or standalone

jesse-greathouse committed:
Merge branch 'update/SDS-parsely-plugin-stats' of github.com:Automattic/vip-go-mu-plugins into update/SDS-parsely-plugin-stats

mjangda commented on config/class-site-details-index.php:
```diff
@@ -228,6 +248,74 @@ public function get_jetpack_info() {
+	 */
+	public function get_parsely_integration_type() {
+		// detect a hidden installation
+		if ( false !== get_option( '_wpvip_parsely_mu', false ) && ! has_filter( 'wpvip_parsely_load_mu' ) ) {
```
Rather than using `has_filter` to detect things, I wonder if it would be better to set have the code in `wp-parsely.php` track / share necessary details instead which would we infer the data more directly from.

e.g. could have a static class called `Parsely_Loader_Info` with methods `(get|set)_integration_type`. `wp-parsely.php` would call `Parsely_Loader_Info::set_integration_type( ... )` and here we'd fetch it via `Parsely_Loader_Info::get_integration_type()`.

jesse-greathouse replied:
> Rather than using `has_filter` to detect things, I wonder if it would be better to set have the code in `wp-parsely.php` track / share necessary details instead which would we infer the data more directly from.
> 
> e.g. could have a static class called `Parsely_Loader_Info` with methods `(get|set)_integration_type`. `wp-parsely.php` would call `Parsely_Loader_Info::set_integration_type( ... )` and here we'd fetch it via `Parsely_Loader_Info::get_integration_type()`.

Im not opposed to this change but I'm not clear on what problem it is attempting to solve by this. I'm anxious to hear more of your wisdom on this.

jesse-greathouse replied:
I have moved most of the logic into the new `Parsely_Loader_info` class in `wp-parsely.php`. It has all of the functionality that you've asked for and will keep these values in static properties so that they only have to be resolved once per property.

jesse-greathouse committed:
changed array value: parsely_service_type to service_type

acicovic commented on config/class-site-details-index.php:
```diff
@@ -228,6 +248,74 @@ public function get_jetpack_info() {
+	 */
+	public function get_parsely_integration_type() {
+		// detect a hidden installation
+		if ( false !== get_option( '_wpvip_parsely_mu', false ) && ! has_filter( 'wpvip_parsely_load_mu' ) ) {
```
I think that `false !==` is not needed here?

jesse-greathouse replied:
Thank you for the feedback. I have made this change in the most recent commit.

acicovic commented on config/class-site-details-index.php:
```diff
@@ -228,6 +248,74 @@ public function get_jetpack_info() {
+
+		// is enabled as a standard plugin
+		$plugin = 'wp-parsely/wp-parsely.php';
+		if ( in_array( $plugin, (array) get_option( 'active_plugins', array() ), true ) || is_plugin_active_for_network( $plugin ) ) {
```
Have we considered using [is_plugin_active()](https://developer.wordpress.org/reference/functions/is_plugin_active/) here, or it isn't possible for some reason?

jesse-greathouse replied:
Thank you for the feedback. I'm not sure why I didn't do that initially. I have made the change in the most recent commit.

acicovic committed:
Improve some comments related to Parse.ly

jesse-greathouse committed:
* Moved most logic into new Parsely_Loader_Info class in wp-parsely.php
* Removed unnecessary false !== in set_integration_type method
* Consolodated logic into is_plugin_active in set_integration_type method

jesse-greathouse committed:
* fixed merge conflicts

jesse-greathouse committed:
Merge branch 'develop' into update/SDS-parsely-plugin-stats

jesse-greathouse committed:
Replaced the comment updates by @acicovic which got lost when I migrated the code to a different file

jesse-greathouse committed:
Merge branch 'update/SDS-parsely-plugin-stats' of github.com:Automattic/vip-go-mu-plugins into update/SDS-parsely-plugin-stats

jesse-greathouse committed:
fixed uninitialized value bug

jesse-greathouse committed:
comment fixes

jesse-greathouse committed:
fixed constant `VERSION_NONE`

jesse-greathouse committed:
set a default value in Parsely_Loader_Info::set_parsely so that the property would always correctly be an array

jesse-greathouse committed:
changed isset to array_key_exists in set_version

jesse-greathouse committed:
flipped logic in Parsely_Loader_Info::set_version

acicovic commented on config/class-site-details-index.php:
```diff
@@ -228,6 +231,20 @@ public function get_jetpack_info() {
+		$parsely_info['active']           = Pinfo::get_active();
+		$parsely_info['integration_type'] = Pinfo::get_integration_type();
+		$parsely_info['service_type']     = Pinfo::get_service_type();
+		$parsely_info['verison']          = Pinfo::get_version();
```
```suggestion
		$parsely_info['version']          = Pinfo::get_version();
```

acicovic commented on config/class-site-details-index.php:
```diff
@@ -228,6 +231,20 @@ public function get_jetpack_info() {
 	}
 
+	/**
+	 * Gather all the information about Parsely
```
```suggestion
	 * Gather all the information about Parse.ly.
```

acicovic commented on config/class-site-details-index.php:
```diff
@@ -228,6 +231,20 @@ public function get_jetpack_info() {
 
+	/**
+	 * Gather all the information about Parsely
+	 * @return array Parsely plugin info.
```
```suggestion
	 * @return array Parse.ly plugin info.
```

acicovic commented on wp-parsely.php:
```diff
@@ -21,6 +21,189 @@
 ];
 
+/**
+ * Class to hold Parse.ly loading info
```
```suggestion
 * Class to hold Parse.ly loading info.
```

acicovic commented on wp-parsely.php:
```diff
@@ -21,6 +21,189 @@
+/**
+ * Class to hold Parse.ly loading info
+ * Will prevent this information from being queried multiple times
+ * by keeping the status in a static variable
```
```suggestion
 * by keeping the status in a static variable.
```

acicovic commented on wp-parsely.php:
```diff
@@ -21,6 +21,189 @@
+	private static $version;
+
+	/**
+	 * fetches the active status.
```
```suggestion
	 * Fetches the active status.
```

acicovic commented on wp-parsely.php:
```diff
@@ -21,6 +21,189 @@
+	}
+
+	/**
+	 * populates the private $active property if it hasn't been set.
```
```suggestion
	 * Populates the private $active property if it hasn't been set.
```

acicovic commented on wp-parsely.php:
```diff
@@ -21,6 +21,189 @@
+	}
+
+	/**
+	 * fetches the integration type.
```
```suggestion
	 * Fetches the integration type.
```

acicovic commented on wp-parsely.php:
```diff
@@ -21,6 +21,189 @@
+	}
+
+	/**
+	 * populates the private $integration_type property if it hasn't been set.
```
```suggestion
	 * Populates the private $integration_type property if it hasn't been set.
```

acicovic commented on wp-parsely.php:
```diff
@@ -21,6 +21,189 @@
+			self::$integration_type = self::INTEGRATION_TYPE_MUPLUGINS;
+		}
+
+		// is enabled as a standard plugin
```
```suggestion
		// Is enabled as a standard plugin.
```

acicovic commented on wp-parsely.php:
```diff
@@ -21,6 +21,189 @@
+	}
+
+	/**
+	 * fetches Parse.ly options dictionary.
```
```suggestion
	 * Fetches Parse.ly options dictionary.
```

acicovic commented on wp-parsely.php:
```diff
@@ -21,6 +21,189 @@
+	}
+
+	/**
+	 * populates the private $parsely property if it hasn't been set yet.
```
```suggestion
	 * Populates the private $parsely property if it hasn't been set yet.
```

acicovic commented on wp-parsely.php:
```diff
@@ -21,6 +21,189 @@
+	}
+
+	/**
+	 * fetches the service type.
```
```suggestion
	 * Fetches the service type.
```

acicovic commented on wp-parsely.php:
```diff
@@ -21,6 +21,189 @@
+	}
+
+	/**
+	 * populates the private $service_type property if it hasn't been set.
```
```suggestion
	 * Populates the private $service_type property if it hasn't been set.
```

acicovic commented on wp-parsely.php:
```diff
@@ -21,6 +21,189 @@
+	}
+
+	/**
+	 * fetches the version.
```
```suggestion
	 * Fetches the version.
```

acicovic commented on wp-parsely.php:
```diff
@@ -21,6 +21,189 @@
+	}
+
+	/**
+	 * populates the private $version property if it hasn't been set.
```
```suggestion
	 * Populates the private $version property if it hasn't been set.
```

acicovic reviewed (commented):
Left some spelling suggestions.

acicovic commented on wp-parsely.php:
```diff
@@ -21,6 +21,189 @@
+	 * populates the private $integration_type property if it hasn't been set.
+	 */
+	private static function set_integration_type() {
+		// detect a hidden installation
```
```suggestion
		// Detect a hidden installation.
```

acicovic commented on wp-parsely.php:
```diff
@@ -21,6 +21,189 @@
+			self::$integration_type = self::INTEGRATION_TYPE_MUPLUGINS_SILENT;
+		}
+
+		// mu-plugins
```
```suggestion
		// Is enabled by mu-plugins.
```

acicovic commented on wp-parsely.php:
```diff
@@ -21,6 +21,189 @@
+		}
+
+		// mu-plugins
+		if ( has_filter( 'wpvip_parsely_load_mu' ) ) {
```
What happens here if the `wpvip_parsely_load_mu` filter is set to `false`, and would there be any unwanted side-effects to the `set_active()` function? Is there a fix that must be applied here or in the `set_active()` function?

jesse-greathouse replied:
Is there a way to tell what values are present in an applied filter?

acicovic replied:
I would think there is, but a quick search doesn't reveal anything. I can look into that more tomorrow if needed.

acicovic replied:
Maybe what we could do (and maybe what @mjangda was suggesting?) is to change the `Parsely_Loader_Info` state from within the `maybe_load_plugin()` and `maybe_disable_some_features()` functions where the actual operations take place? @jesse-greathouse, can you give it a shot and see if this approach can work?

If the answer is yes, I can help you out more if needed.

jesse-greathouse replied:
@acicovic,

The latest commit is trying to use this suggestion and I think it works well. The commit notes are as follows:

```
Changed Parsely_Loader_Info setter methods to public
Gave Parsely_Loader_Info setter methods a parameter to pass the value
Integreated Parsely_Loader_Info with maybe_load_plugin and maybe_disable_some_features methods in wp-parsely.php
Added appropriate docblock metadata for the new parameters in setters
Removed unnecessary code in Parsely_Loader_Info methods
```

If you do not mind, please look over the way I used the setter methods in `maybe_load_plugin` and `maybe_disable_some_features` to be sure that it is correctly setting the appropriate values. If I messed it up, it should be easy to fix.

acicovic commented on wp-parsely.php:
```diff
@@ -21,6 +21,189 @@
+	/**
+	 * @var array The Parse.ly WordPress options dictionary.
+	 */
+	private static $parsely;
```
For clarity, I would suggest to rename this variable and related functions to `*parsely_option` or similar.

jesse-greathouse replied:
You are right. This should be renamed. I have refactored this property, as well as its getter/setter methods, in the latest commit.

acicovic commented on wp-parsely.php:
```diff
@@ -21,6 +21,189 @@
+		}
+
+		self::$service_type = self::SERVICE_TYPE_NONE;
+
```
This empty line (178) can be removed.

jesse-greathouse replied:
Thank you. I have corrected this in the latest commit.

acicovic commented on wp-parsely.php:
```diff
@@ -21,6 +21,189 @@
+
+		self::$version = self::VERSION_NONE;
+	}
+
```
This empty line (204) can be removed.

jesse-greathouse replied:
Very good. I have changed this in the latest commit.

acicovic commented on wp-parsely.php:
```diff
@@ -21,6 +21,189 @@
+ * by keeping the status in a static variable
+ */
+final class Parsely_Loader_Info {
+
```
This empty line (30) can be removed.

jesse-greathouse replied:
Great! I have removed this line in the latest commit.

acicovic commented:
For the functions within `Parsely_Loader_Info` that return something, you might want to consider adding `@return` tags in the function comments.

jesse-greathouse committed:
Update config/class-site-details-index.php

fixed typo in version

Co-authored-by: Alex Cicovic <23142906+acicovic@users.noreply.github.com>

jesse-greathouse committed:
Update config/class-site-details-index.php

Fixed typo.

Co-authored-by: Alex Cicovic <23142906+acicovic@users.noreply.github.com>

jesse-greathouse committed:
Update config/class-site-details-index.php

Fixed typo.

Co-authored-by: Alex Cicovic <23142906+acicovic@users.noreply.github.com>

jesse-greathouse committed:
Update wp-parsely.php

Fixed typo.

Co-authored-by: Alex Cicovic <23142906+acicovic@users.noreply.github.com>

jesse-greathouse committed:
Update wp-parsely.php

Fixed typo.

Co-authored-by: Alex Cicovic <23142906+acicovic@users.noreply.github.com>

jesse-greathouse committed:
Update wp-parsely.php

Fixed typo.

Co-authored-by: Alex Cicovic <23142906+acicovic@users.noreply.github.com>

jesse-greathouse committed:
Update wp-parsely.php

Fixed typo.

Co-authored-by: Alex Cicovic <23142906+acicovic@users.noreply.github.com>

jesse-greathouse committed:
Update wp-parsely.php

Fixed typo.

Co-authored-by: Alex Cicovic <23142906+acicovic@users.noreply.github.com>

jesse-greathouse committed:
Update wp-parsely.php

Fixed typo.

Co-authored-by: Alex Cicovic <23142906+acicovic@users.noreply.github.com>

jesse-greathouse committed:
Update wp-parsely.php

Fixed typo.

Co-authored-by: Alex Cicovic <23142906+acicovic@users.noreply.github.com>

jesse-greathouse committed:
Update wp-parsely.php

Fixed typo.

Co-authored-by: Alex Cicovic <23142906+acicovic@users.noreply.github.com>

jesse-greathouse committed:
Update wp-parsely.php

Fixed typo.

Co-authored-by: Alex Cicovic <23142906+acicovic@users.noreply.github.com>

jesse-greathouse committed:
Update wp-parsely.php

Fixed typo.

Co-authored-by: Alex Cicovic <23142906+acicovic@users.noreply.github.com>

jesse-greathouse committed:
Update wp-parsely.php

Fixed typo.

Co-authored-by: Alex Cicovic <23142906+acicovic@users.noreply.github.com>

jesse-greathouse committed:
Update wp-parsely.php

Fixed typo.

Co-authored-by: Alex Cicovic <23142906+acicovic@users.noreply.github.com>

jesse-greathouse committed:
* changed property `parsely` to `parsely_options`
* changed method `get_parsely` to `get_parsely_options`
* changed method `set_parsely` to `set_parsely_options`
* removed empty line (178)
* removed empty line (204)
* removed empty line (30)
* Added `@return` annotations to docblocks

jesse-greathouse commented:
> For the functions within `Parsely_Loader_Info` that return something, you might want to consider adding `@return` tags in the function comments.

Thank you. This has been added in the latest commit.

jesse-greathouse committed:
Merge branch 'develop' into update/SDS-parsely-plugin-stats

jesse-greathouse committed:
Update wp-parsely.php

Capitalize Populates.

Co-authored-by: Alex Cicovic <23142906+acicovic@users.noreply.github.com>

jesse-greathouse committed:
Update wp-parsely.php

Corrected Spelling.

Co-authored-by: Alex Cicovic <23142906+acicovic@users.noreply.github.com>

jesse-greathouse committed:
Update wp-parsely.php

Spelling

Co-authored-by: Alex Cicovic <23142906+acicovic@users.noreply.github.com>

jesse-greathouse committed:
Merge branch 'develop' into update/SDS-parsely-plugin-stats

jesse-greathouse committed:
Changed Parsely_Loader_Info setter methods to public
Gave Parsely_Loader_Info setter methods a parameter to pass the value
Integreated Parsely_Loader_Info with maybe_load_plugin and maybe_disable_some_features methods in wp-parsely.php
Added appropriate docblock metadata for the new parameters in setters
Removed unnecessary code in Parsely_Loader_Info methods

acicovic committed:
wp-parsely.php: Update for stricter Loader Info defaults and readings accuracy

acicovic committed:
wp-parsely.php: Remove "PAID" and "NONE" service types and default to "UNKNOWN".

acicovic committed:
wp-parsely.php: Replace "VERSION_NONE" with "VERSION_UNKNOWN"

jesse-greathouse committed:
Changes that will allow the user to activate wp-parsely in the wp-admin interface

jesse-greathouse committed:
fixed bug where using or instead of and

jesse-greathouse committed:
Added logic for detecting the version when the plugin is in a self-managed state
* Added semantic variable changes for `$plugin_is_already_loaded` and `$plugin_loading_queued`

acicovic commented on wp-parsely.php:
```diff
@@ -47,12 +226,44 @@ function maybe_load_plugin() {
+		}
+
+		// Set Loader Info data as an active self-managed integration.
+		if ( $plugin_is_already_loaded || $plugin_loading_queued ) {
+			Parsely_Loader_Info::set_active( true );
+			Parsely_Loader_Info::set_integration_type( Parsely_Loader_Info::INTEGRATION_TYPE_SELF_MANAGED );
+
+			//Get the version of the self-managed wp-parsely plugin.
+			$parsely_options = Parsely_Loader_Info::get_parsely_options();
+			if ( array_key_exists( 'plugin_version', $parsely_options ) ) {
+				Parsely_Loader_Info::set_version( $parsely_options['plugin_version'] );
+			}
+		}
```
Question: I'm wondering whether we want to run this during plugin activation? Opinions?

jesse-greathouse replied:
I don't have strong opinions on this. I've removed it in the latest commit.

acicovic commented:
I'm ready to approve this PR when [this](https://github.com/Automattic/vip-go-mu-plugins/pull/3542#discussion_r957059149) gets sorted out,but I'd need feedback from other reviewers as well. Thanks!

jesse-greathouse committed:
Values will not be collected while plugin is being activated

jesse-greathouse committed:
Merge branch 'develop' into update/SDS-parsely-plugin-stats

jesse-greathouse committed:
test coverage for new code

jesse-greathouse committed:
Merge branch 'update/SDS-parsely-plugin-stats' of github.com:Automattic/vip-go-mu-plugins into update/SDS-parsely-plugin-stats

jesse-greathouse committed:
fixing tests

jesse-greathouse committed:
removed test that tested defaults because the test always gets ran after the parsely options have been set

jesse-greathouse committed:
added back in the test for defaults, which conditionally only tests for actual defaults in "disabled" mode

chriszarate commented on wp-parsely.php:
```diff
@@ -47,12 +226,44 @@ function maybe_load_plugin() {
+
+	// Check if wp-parsely has already been loaded, for example when the user
+	// has a self-managed wp-parsely installation.
+	$plugin_is_already_loaded = in_array( PARSELY_PLUGIN_SIGNATURE, get_option( 'active_plugins' ) );
```
guard against missing option
```suggestion
	$plugin_is_already_loaded = in_array( PARSELY_PLUGIN_SIGNATURE, get_option( 'active_plugins', [] ), true );
```

jesse-greathouse replied:
Thank you. This has been corrected in the most recent commit.

chriszarate commented on wp-parsely.php:
```diff
@@ -47,12 +226,44 @@ function maybe_load_plugin() {
-	if ( ! apply_filters( 'wpvip_parsely_load_mu', get_option( '_wpvip_parsely_mu' ) === '1' ) ) {
-		return;
-	}
+	$do_not_load_plugin = ! apply_filters( 'wpvip_parsely_load_mu', get_option( '_wpvip_parsely_mu' ) === '1' );
```
having this spelled out and separate is useful

```suggestion
	$load_plugin_requested_via_option = get_option( '_wpvip_parsely_mu' ) === '1';
	$do_not_load_plugin = ! apply_filters( 'wpvip_parsely_load_mu', $load_plugin_requested_via_option );
```

jesse-greathouse replied:
Awesome suggestion. Thanks 👍

chriszarate commented on wp-parsely.php:
```diff
@@ -47,12 +226,44 @@ function maybe_load_plugin() {
+			Parsely_Loader_Info::set_active( false );
+			Parsely_Loader_Info::set_integration_type( Parsely_Loader_Info::INTEGRATION_TYPE_MUPLUGINS );
+
+			if ( get_option( '_wpvip_parsely_mu' ) === '1' ) {
```
```suggestion
			if ( $load_plugin_requested_via_option ) {
```

jesse-greathouse replied:
Very good. I have added this.

chriszarate commented on wp-parsely.php:
```diff
@@ -47,12 +226,44 @@ function maybe_load_plugin() {
+	$plugin_loading_queued = is_queued_for_activation();
+
+	// If the plugin must not be loaded, set Loader Info data and exit.
+	if ( $do_not_load_plugin || $plugin_is_already_loaded || $plugin_loading_queued ) {
```
This is ... incredibly complicated with not enough comments. It also feels like this is in the wrong order. It would benefit from very descriptive variable names and early returns, like:

```
if ( $self_managed_plugin_has_loaded || $self_managed_plugin_is_queued_for_activation ) {
  Parsely_Loader_Info::set_active( true );
  Parsely_Loader_Info::set_integration_type( Parsely_Loader_Info::INTEGRATION_TYPE_SELF_MANAGED );
  return;
}

if ( $load_plugin_requested_via_option && $force_disabled ) {
  Parsely_Loader_Info::set_active( false );
  Parsely_Loader_Info::set_integration_type( Parsely_Loader_Info::INTEGRATION_TYPE_MU_PLUGINS );
  return;
}

// etc
```

jesse-greathouse replied:
Thank you for the excellent suggestions for this very convoluted logic. This section has been one of the most difficult things about working on this piece of code. Just understanding what's happening is a nightmare. Thanks to your suggestion I have made this section a lot less convoluted and more readable, and I've added a lot of relevant comments to help the next person who might have to read this. I feel that the latest commit makes this section a lot better. Thank you very much for your time and great thoughts on this.

jesse-greathouse committed:
Merge branch 'develop' into update/SDS-parsely-plugin-stats

jesse-greathouse committed:
Refactored the maybe_load_plugin method with help from @chriszarate
* Guarded against missing option in `get_option`
* Changed variable names to be more self-documenting
* Refactored the conveluted logic to make it more readable and efficient
* Changed the default wp-parse.ly version in tests because 3.0 has been moved out

jesse-greathouse committed:
fixed comment

jesse-greathouse committed:
Merge branch 'develop' into update/SDS-parsely-plugin-stats

acicovic commented on wp-parsely.php:
```diff
@@ -47,12 +226,43 @@ function maybe_load_plugin() {
+	// has a self-managed wp-parsely installation.
+	$self_managed_plugin_has_loaded = in_array( PARSELY_PLUGIN_SIGNATURE, get_option( 'active_plugins', [] ) );
+
+	if ( $self_managed_plugin_has_loaded || $self_managed_plugin_is_queued_for_activation ) {
+		$parsely_options = Parsely_Loader_Info::get_parsely_options();
+		Parsely_Loader_Info::set_active( true );
```
My only remaining question here is whether we should execute this for `$self_managed_plugin_is_queued_for_activation`, instead of just returning. @chriszarate, @mjangda, opinions?

jesse-greathouse replied:
> My only remaining question here is whether we should execute this for `$self_managed_plugin_is_queued_for_activation`, instead of just returning. @chriszarate, @mjangda, opinions?

Adding @rinatkhaziev and @WPprodigy. They may have some thoughts on the question by @acicovic

acicovic replied:
Thanks!

WPprodigy replied:
All the `is_queued_for_activation()` related code paths (and tests) feels like it's adding more complication than is worth IMO. Can we just strip that part out? I don't _think_ we really need to know right that second on the activation requests themselves. The next PHP request will know about it's activated/deactivation, and we're never going to be activating & syncing sds data on the same request.

jesse-greathouse replied:
This part was not added to record the info for SDS. `is_queued_for_activation()` was added because when a user activates a self managed Parse.ly plugin installation, and the code has `add_filter( 'wpvip_parsely_load_mu', '__return_true' );` somewhere in their code, there will be a class loading collision between the mu-plugins integrated wp-parsely, and their self-managed installation.
![Screen Shot 2022-08-26 at 9 21 12 AM](https://user-images.githubusercontent.com/46167019/188753159-f1016caa-dbd7-476f-b358-8c77fce63802.png)

jesse-greathouse replied:
> ...and we're never going to be activating & syncing sds data on the same request.

@acicovic, I think this is the answer we've been seeking on your question. Based on @WPprodigy's input, there will not be a need to populate this data for use in the SDS request, because there will not be a sync to SDS on this request. I will make the necessary change that will short circuit this logic when activation is queued on this request.

jesse-greathouse replied:
I have addressed this in my latest commit. [The SDS info will not be captured when the request has an activation queued](https://github.com/Automattic/vip-go-mu-plugins/pull/3542/files#diff-41f0c604e0848ecb9ac1b6e07d5d2fab90b2ef3f715792c15243c5eed180894cR249).

acicovic replied:
Great, thank you both!

jesse-greathouse committed:
Merge branch 'develop' into update/SDS-parsely-plugin-stats

jesse-greathouse committed:
Merge branch 'develop' into update/SDS-parsely-plugin-stats

WPprodigy committed:
Merge branch 'develop' into update/SDS-parsely-plugin-stats

WPprodigy commented on wp-parsely.php:
```diff
@@ -47,12 +226,43 @@ function maybe_load_plugin() {
+	// The activation of this plugin has been queued in this request.
+	$self_managed_plugin_is_queued_for_activation = is_queued_for_activation();
+
+	// Check if wp-parsely has already been loaded, for example when the user
+	// has a self-managed wp-parsely installation.
+	$self_managed_plugin_has_loaded = in_array( PARSELY_PLUGIN_SIGNATURE, get_option( 'active_plugins', [] ) );
```
This isn't a safe enough protection to use to abort early if the plugin is already loaded. The plugin signature does not have to be `wp-parsely/wp-parsely.php`, it could be anything like `wp-parsely-2/wp-parsely.php` for example. Or it could be manually `require`'d in client-mu-plugins, etc.

We need to return to the safety of `class_exists( 'Parsely' ) || class_exists( 'Parsely\Parsely' )` type checks.

jesse-greathouse replied:
In my testing, `class_exists( 'Parsely' ) || class_exists( 'Parsely\Parsely' )` never actually worked when the user was using a self-managed plugin. I believe it is because this code is executed before the plugin was loaded.

jesse-greathouse replied:
Please see the changes in my latest commit which address all of the concerns that you brought up. I have tested the scenarios and it seems to be all just fine. 

The `class_exists( 'Parsely' ) || class_exists( 'Parsely\Parsely' )` logic does work as expected when a user is doing a direct require in `client-mu-plugins/plugin-loader.php` or putting their wp-parsely directly in `client-mu-plugins`.

If this class is not found, it reverse to the methodology of checking the `active_plugins` list, but this time it uses a regex pattern to identify anything that looks like it's a wp-parsely plugin. e.g. `wp-parsely2/wp-parsely.php`, `wp-parsely-1.10/wp-parsely.php`, etc...

I believe that if there is any more of a wild Parse.ly plugin out there, then this would be such a small statistical anomaly that it's not really all that relevant for this scope.

Here is the commit log:

* Added back in `class_exists( 'Parsely' ) || class_exists( 'Parsely\Parsely' )` logic to account for when the plugin is loaded through `client-mu-plugins`.
* Added function: `has_plugin_signature` which uses a regex pattern to determine if a plugin is loaded that looks like wp-parsely.
* Changed constant `PARSELY_PLUGIN_SIGNATURE` to `PARSELY_PLUGIN_SIGNATURE_REGEX` to represent the pattern which a wp-parsely plugin could look like.
* Updated tests and added tests for `has_plugin_signature`
* Thoroughly tested all scenarios so that I know it works.

jesse-greathouse committed:
Merge branch 'develop' into update/SDS-parsely-plugin-stats

jesse-greathouse committed:
Significant refactor due to issues brought up by @WPprodigy
  * Added back in `class_exists( 'Parsely' ) || class_exists( 'Parsely\Parsely' )` logic to account for when the plugin is loaded through `client-mu-plugins`.
  * Added function: `has_plugin_signature` which uses a regex pattern to determine if a plugin is loaded that looks like wp-parsely.
  * Changed constant `PARSELY_PLUGIN_SIGNATURE` to `PARSELY_PLUGIN_SIGNATURE_REGEX` to represent the pattern which a wp-parsely plugin could look like.
  * Updated tests and added tests for `has_plugin_signature`
  * Thoroughly tested all scenarios so that I know it works.

jesse-greathouse committed:
Merge branch 'develop' into update/SDS-parsely-plugin-stats

jesse-greathouse committed:
Merge branch 'update/SDS-parsely-plugin-stats' of github.com:Automattic/vip-go-mu-plugins into update/SDS-parsely-plugin-stats

jesse-greathouse committed:
code quality

jesse-greathouse committed:
code quality

jesse-greathouse committed:
fixed tests

jesse-greathouse committed:
sanitize inputs

WPprodigy committed:
possible refactor

Parsely_Loader_Info:
- use types in the method signatures
- check uninitialized properties with isset() instead of comparing to null (avoids php warnings)
- Rather than setting “unknown”s, just return it as a default.

General:
- cleanup is_queued_for_activation(), a bit more strict on when it will run now.
- simplify the filter & option enabled/disable functionality
- load on plugins_loaded at priority 1, this way we catch the plugin already being activated via the UI without having to do imperfect regex things.

Questions:
- Why is get_service_type/set_service_type needed?

jesse-greathouse committed:
Merge branch 'develop' into update/SDS-parsely-plugin-stats

rinatkhaziev commented:
@jesse-greathouse Parse.ly-related tests are failing.

jesse-greathouse committed:
Merge branch 'develop' into update/SDS-parsely-plugin-stats

jesse-greathouse committed:
maybe_fix_tests()

jesse-greathouse committed:
Merge branch 'update/SDS-parsely-plugin-stats' of github.com:Automattic/vip-go-mu-plugins into update/SDS-parsely-plugin-stats

jesse-greathouse commented:
> @jesse-greathouse Parse.ly-related tests are failing.

Tnx, @rinatkhaziev. I fixed the tests in the latest commit.

jesse-greathouse committed:
Merge branch 'develop' into update/SDS-parsely-plugin-stats

WPprodigy committed:
Merge branch 'develop' into update/SDS-parsely-plugin-stats

WPprodigy commented on wp-parsely.php:
```diff
@@ -21,6 +21,67 @@
+		self::$integration_type = $integration_type;
+	}
+
+	public static function get_service_type(): string {
+		return isset( self::$service_type ) ? self::$service_type : self::SERVICE_TYPE_UNKNOWN;
+	}
+
+	public static function set_service_type( string $service_type ): void {
+		self::$service_type = $service_type;
+	}
```
Should we strip these out until/if we need them? Not a big fan of shipping unused "maybe we use this in the future" type code. Always easier to add than to remove later :)

acicovic replied:
IMO we can strip these out.

jesse-greathouse replied:
I have removed this in the latest commit.

jesse-greathouse replied:
I will mirror that change in the SDS code.

WPprodigy reviewed (dismissed):
I updated the PR description to account for what all is changed here.

Also did a bunch of testing and all looks good as far as I can tell! Used this in my testing as I went through the various states:

```
add_action( 'init', function() {
	$parsely_info                     = [];
	$parsely_info['active']           = ParselyInfo::is_active();
	$parsely_info['integration_type'] = ParselyInfo::get_integration_type();
	$parsely_info['service_type']     = ParselyInfo::get_service_type();
	$parsely_info['version']          = ParselyInfo::get_version();
	error_log( print_r( $parsely_info, true ) );
} );

// add_filter( 'wpvip_parsely_load_mu', '__return_true' );
// add_filter( 'pre_option__wpvip_parsely_mu', fn() => '1' );
```

And then I confirmed `Automattic\VIP\Config\Site_Details_Index::instance()->get_site_details();` output looks good.

We do need to be vigilant on deploy due to the `plugins_loaded` hook change, but I'm quite sure it should go over smoothly.

jesse-greathouse committed:
Merge branch 'develop' into update/SDS-parsely-plugin-stats

jesse-greathouse committed:
removed service type from Parse.ly info; hopefully to be added back later when it can be implemented

rinatkhaziev commented:
> I'm quite sure it should go over smoothly.

Famous last words :)

We should check/test against customers who load the standalone versions prior to the release.

WPprodigy commented:
Can definitely check/test those to be safe. More details on how it could break:

If a self-installed version is loaded via UI/wpcom_vip_load_plugin(), all is fine. The only time it could go poorly is if the plugin file is manually require'd on a hook later than `plugins_loaded` priority 1 - but doing so would have caused the previous (current) code to fatal out anyway though (unless using the filter to disable).

For things to break that weren't currently broken, you'd have to use the `muplugins_loaded` hook to see if we loaded parsely, and then use the result of that check to manually require the plugin on `plugins_loaded` after priority 1. Lots of random hoops to jump through in order to make this change not backwards-compat. All in all, extremely unlikely bordering impossible.

The other scenario that would change is unhooking the `maybe_load_plugin` hook - but we've documented using the `wpvip_parsely_load_mu ` filter instead of that route.

jesse-greathouse committed:
Merge branch 'develop' into update/SDS-parsely-plugin-stats

jesse-greathouse committed:
Merge branch 'develop' into update/SDS-parsely-plugin-stats
